storage: reapply the rule solver

BramGruneir opened the pull request:
This adds back in 3 commits that were removed to facilitate the merge of develop
back to master. One other commit, is no longer required.

Closes #9336

1) 4446345bf7a9e1a6eb7c7a3c71ec9c8e3b3d17fc
storage: add constraint rule solver for allocation

Rules are represented as a single function that returns the candidacy of the
store as well as a float value representing the score. These scores are then
aggregated from all rules and returns the stores sorted by them.

Current rules:
- ruleReplicasUniqueNodes ensures that no two replicas are put on the same node.
- ruleConstraints enforces that required and prohibited constraints are
  followed, and that stores with more positive constraints are ranked higher.
- ruleDiversity ensures that nodes that have the fewest locality tiers in common
  are given higher priority.
- ruleCapacity prioritizes placing data on empty nodes when the choice is
  available and prevents data from going onto mostly full nodes.

2) dd3229a718d5e3ec25a784f2f3ff5c9ffd85a3b0
storage: implemented RuleSolver into allocator

3) 27353a8d4d90406cb5ff21d4f8198b7e5b5e2325
storage: removed unused rangeCountBalancer

There was a 4th commit that is no longer required. The simulation was already
converging since adding a rebalance threshold.
4e29a36e7806a26ad4d7b5c1b3c06037d328340b
storage/simulation: only rebalance 50% of ranges on each iteration so it will
converge

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10252)

<!-- Reviewable:end -->

BramGruneir committed:
storage: reapply the rule solver

This adds back in 3 commits that were removed to facilitate the merge of develop
back to master. One other commit, is no longer required.

Follow up fixes are tracked in #10275.

Closes #9336

1) 4446345bf7a9e1a6eb7c7a3c71ec9c8e3b3d17fc
storage: add constraint rule solver for allocation

Rules are represented as a single function that returns the candidacy of the
store as well as a float value representing the score. These scores are then
aggregated from all rules and returns the stores sorted by them.

Current rules:
- ruleReplicasUniqueNodes ensures that no two replicas are put on the same node.
- ruleConstraints enforces that required and prohibited constraints are
  followed, and that stores with more positive constraints are ranked higher.
- ruleDiversity ensures that nodes that have the fewest locality tiers in common
  are given higher priority.
- ruleCapacity prioritizes placing data on empty nodes when the choice is
  available and prevents data from going onto mostly full nodes.

2) dd3229a718d5e3ec25a784f2f3ff5c9ffd85a3b0
storage: implemented RuleSolver into allocator

3) 27353a8d4d90406cb5ff21d4f8198b7e5b5e2325
storage: removed unused rangeCountBalancer

There was a 4th commit that is no longer required. The simulation was already
converging since adding a rebalance threshold.
4e29a36e7806a26ad4d7b5c1b3c06037d328340b
storage/simulation: only rebalance 50% of ranges on each iteration so it will
converge

tamird commented:
Reviewed 12 of 14 files at r1.
Review status: 12 of 14 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5B5X-r5F4buO3kwzP:-KV5B5X-r5F4buO3kwzQ:btebsw2) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L115)):*
> ```Go
> 	storePool  *StorePool
> 	options    AllocatorOptions
> 	ruleSolver *ruleSolver
> ```

this can be a value

---

*[pkg/storage/allocator.go, line 174 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5BEU8DB6ZDWXY-KTy:-KV5BEU8DB6ZDWXY-KTz:bwu5fw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L174)):*
> ```Go
> func (a *Allocator) AllocateTarget(
> 	constraints config.Constraints, existing []roachpb.ReplicaDescriptor, rangeID roachpb.RangeID,
> ) (*roachpb.StoreDescriptor, error) {
> ```

hm, seems like this wants to return a value now instead of a pointer

---

*[pkg/storage/allocator.go, line 185 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5BLBfMmoWhW7y4Rpc:-KV5BLBfMmoWhW7y4Rpd:boxvz5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L185)):*
> ```Go
> 		}
> 	}
> 	return &candidates[0].store, nil
> ```

how come the randomness that was previously here is no longer needed?

---

*[pkg/storage/allocator.go, line 201 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5BYIA_c5GRHSQhXkl:-KV5BYIA_c5GRHSQhXkm:bgdgnpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L201)):*
> ```Go
> 	}
> 
> 	// TODO(bram): Is this getStoreList call required?
> ```

expand this rationale; why is it currently required, why might it not be required? why is it OK to use a bogus range ID?

---

*[pkg/storage/allocator.go, line 207 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5ElRI1pw1kbbooWAz:-KV5ElRI1pw1kbbooWB-:bbmfjxe) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L207)):*
> ```Go
> 	found := false
> 	var worst roachpb.ReplicaDescriptor
> 	var worstScore float64
> ```

seems like you can initialize this to `math.MaxFloat64` and remove `found`.

---

*[pkg/storage/allocator.go, line 217 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5EiAqh5ZTwp0j2l7z:-KV5EiAqh5ZTwp0j2l8-:b-tnpz4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L217)):*
> ```Go
> 		}
> 		// If it's not a valid candidate, score will be zero.
> 		candidate, _ := a.ruleSolver.computeCandidate(solveState{
> ```

why is it OK to skip checking the error?

---

*[pkg/storage/allocator.go, line 269 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5F276RijvKQ6OxzcR:-KV5F276RijvKQ6OxzcS:bzic3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L269)):*
> ```Go
> 	}
> 
> 	// TODO(bram): getStoreList is called twice, once now, and again in solve,
> ```

`getStoreList` is also the only function on `storePool` that is ever called from the solver, and that function (`Solve()`) is only ever called from the allocator, which can pass the store list.

Seems easy enough to fix in this PR by passing the store list as an argument to Solve; WDYT?

---

*[pkg/storage/allocator.go, line 299 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5GLJXHMbI5kAoE-RY:-KV5GLJXHMbI5kAoE-RZ:b-fwk9mm) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L299)):*
> ```Go
> 	// Find a candidate that is better than one of the existing stores, otherwise
> 	// return nil.
> 	candidatesFound := 0
> ```

I had to squint to figure out what this counter was doing. How about this instead:
```
if len(candidates) > len(existing) {
  candidates = candidates[:len(existing)]
}
```

---

*[pkg/storage/allocator_test.go, line 674 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5Gp3dZhRSUOEkAgzy:-KV5Gp3dZhRSUOEkAgzz:bagm16f) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator_test.go#L674)):*
> ```Go
> 
> 		a.storePool.mu.Lock()
> 		a.storePool.mu.deterministic = true
> ```

doesn't this want to be inside `createTestAllocator`?

---

*[pkg/storage/allocator_test.go, line 1435 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5H1PiQoAWvQpdl15m:-KV5H1PiQoAWvQpdl15n:beylv6e) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator_test.go#L1435)):*
> ```Go
> 
> 	// Output:
> 	// generation    0:   1 88%,  0  0%,  0  0%,  0  0%,  0  0%,  0  0%,  0  0%,  0  0%,  0  0%,  0  0%,  0  0%,  0  0%,  0  0%,  0  0%,  0  0%,  0  0%,  0  0%,  0  0%,  0  0%  1 11%
> ```

please revert this, or put it in a separate commit. This test is important for validating that the allocator's behaviour didn't inadvertently change, and including this graphical change in the rest of the change makes it difficult to spot interesting changes.

---

*[pkg/storage/client_raft_test.go, line 1917 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5Dub6Ahg1EL0Mcodz:-KV5Dub6Ahg1EL0Mcoe-:b-adv9lq) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/client_raft_test.go#L1917)):*
> ```Go
> 		actual := countReplicas()
> 		if !reflect.DeepEqual(expected, actual) {
> 			return errors.Errorf("replicas are not distributed as expected = %+v, want %+v, %s", actual, expected, pretty.Diff(expected, actual))
> ```

why'd this change? the diff alone isn't enough?

---

*[pkg/storage/store_pool.go, line 216 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5DUXc1K6Piy_gSsh1:-KV5DUXc1K6Piy_gSsh2:b-o0lvwl) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/store_pool.go#L216)):*
> ```Go
> 		storeDetails  map[roachpb.StoreID]*storeDetail
> 		queue         storePoolPQ
> 		deterministic bool
> ```

this doesn't need to be under lock, does it?

---

*[pkg/storage/store_pool_test.go, line 42 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5DeBlR1A5TGBlC3xj:-KV5DeBlR1A5TGBlC3xk:b-m8il2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/store_pool_test.go#L42)):*
> ```Go
> 
> // TestSetDeterministic makes StorePool return results in a deterministic way.
> func (sp *StorePool) TestSetDeterministic(deterministic bool) {
> ```

this has one caller.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 12 of 14 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 174 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5BEU8DB6ZDWXY-KTy:-KV5UkwZ0gww4Z735kEd:b-vext2w) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L174)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> hm, seems like this wants to return a value now instead of a pointer
</details>
Let's not change this. Returning values isn't always a win. 

---

*[pkg/storage/allocator.go, line 185 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5BLBfMmoWhW7y4Rpc:-KV5Usi2Bh_6tnjNstHO:b-42tmgp) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L185)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how come the randomness that was previously here is no longer needed?
</details>
+1. The returned candidates are sorted by score which makes this deterministic which is Very Bad.

---

*[pkg/storage/allocator.go, line 310 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5YHbU2nVulMoI0Z-s:-KV5YHbU2nVulMoI0Z-t:bpuswnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L310)):*
> ```Go
> 		}
> 		if !found {
> 			return &store, nil
> ```

We need randomness here too. See my other comment about the candidates being sorted by score at this point making this loop deterministic.

---

*[pkg/storage/store_pool.go, line 518 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5WqWBDmucaEvwgmuK:-KV5WqWBDmucaEvwgmuL:bdf1o1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/store_pool.go#L518)):*
> ```Go
> 	if sp.mu.deterministic {
> 		sort.Sort(sl)
> 	}
> ```

Relying on the map iteration to randomize the `StoreList` is subtle. We should shuffle if `!sp.mu.deterministic`. See `ReplicaSlice.Suffle`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252)*
<!-- Sent from Reviewable.io -->

petermattis commented:
This change is significant enough that I think we want to enable the new rule-based allocator via an env var and have it disabled by default. 

---

Review status: 12 of 14 files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---

*[pkg/storage/rule_solver.go, line 81 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV6YeXjELWyKGtR5kbO:-KV6YeXk3Jk_MEddVZRJ:b-3zsqci) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/rule_solver.go#L81)):*
> ```Go
> // which they are run. For optimization purposes, less computationally intense
> // rules should run first to eliminate candidates.
> func makeRuleSolver(storePool *StorePool, rules []rule) *ruleSolver {
> ```

The generality of a configurable list of rules feels like overkill to me.

---

*[pkg/storage/rule_solver.go, line 122 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV6XcTqVlI3w-r8p9ok:-KV6XcTqVlI3w-r8p9ol:b-uu08bo) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/rule_solver.go#L122)):*
> ```Go
> 	}
> 	sort.Sort(byScore(candidates))
> 	return candidates, nil
> ```

Does any caller every use the list? Might be better to return a single candidate that is randomly selected here. 

One thought on reintroducing randomness: shuffle `sl` and iterate over `sl.stores` until you find two potential candidates and then choose the one with the better score. This be a power-of-two-choices style selection (https://www.eecs.harvard.edu/~michaelm/postscripts/mythesis.pdf).

---

*[pkg/storage/rule_solver.go, line 127 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV6X4xn8aWF55Hy2iyl:-KV6X4xoDk7_F8XtXV0w:ba5eq40) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/rule_solver.go#L127)):*
> ```Go
> // computeCandidate runs all the rules for the store and returns the candidacy
> // information. Returns false if not a candidate.
> func (rs *ruleSolver) computeCandidate(state solveState) (candidate, bool) {
> ```

It is super confusing that `candidate` is a type and we see `candidate, bool` here while in `ruleReplicasUniqueNodes` we see `candidate bool`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-:-KV6ZCL2f9tIgPXYa9dj:bvtpc7f)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Oh, I like that idea a lot.  Going to take a bit of plumbing since this touches a lot.  But I think it's doable.

And a note on this PR.  My goal here is to bring back the code from the original commits more than entirely fixing them.  So by adding the env flag, that should allow some of the details (like how it's not correctly random) to be address in follow up PRs.  Otherwise, this PR would just get out of hand.  

So easy fixes are going in, but substantial changes I'm going to reserve for follow-ups. 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-:-KV6gnTrQ5sIRiKNuSED:bkej3o6)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> So easy fixes are going in, but substantial changes I'm going to reserve for follow-ups.

I'd be fine with seeing this go in without any fixes as long as it is disabled by default. 

---

Review status: 12 of 14 files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-:-KV6ipO05kJxTzY9y2pr:befwp2h)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
*[pkg/storage/allocator.go, line 185 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5BLBfMmoWhW7y4Rpc:-KV6txVoD_UuBGVrKkPf:b-f5i6mm) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L185)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> +1. The returned candidates are sorted by score which makes this deterministic which is Very Bad.
</details>
Why is deterministic order "Very Bad"? Isn't the whole idea of ranking by score such that you can return the best possible candidate? If you're randomizing doesn't that defeat the whole point of positive vs required attributes?

Isn't it being deterministic a net win since it's then also much easier to understand what's happening?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: 12 of 14 files reviewed at latest revision, 18 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 217 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5EiAqh5ZTwp0j2l7z:-KV6wonFX6_xvNvcuu-W:bozfhpr) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L217)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why is it OK to skip checking the error?
</details>
That's not an error. That's a bool whether or not the that store is a potential candidate. If it's not a candidate, the score will be zero and thus will be the worst. Just makes the logic simpler.

---

*[pkg/storage/allocator.go, line 269 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5F276RijvKQ6OxzcR:-KV6xp1vjTwNmgRNiPnD:b-ptd5n6) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L269)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `getStoreList` is also the only function on `storePool` that is ever called from the solver, and that function (`Solve()`) is only ever called from the allocator, which can pass the store list.
> 
> Seems easy enough to fix in this PR by passing the store list as an argument to Solve; WDYT?
</details>
Passing in the storePool was done to support viewing this data in the UI. Since getStoreList wasn't exported, the easiest thing was passing in storePool which was accessible from the admin RPC endpoints.

https://github.com/cockroachdb/cockroach/pull/8887/commits/4601e629a69483721dea1bbe0d46eb909f9e690f 

---

*[pkg/storage/rule_solver.go, line 122 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV6XcTqVlI3w-r8p9ok:-KV6vjonxFdigzD-T4tk:b-s3hz6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/rule_solver.go#L122)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Does any caller every use the list? Might be better to return a single candidate that is randomly selected here. 
> 
> One thought on reintroducing randomness: shuffle `sl` and iterate over `sl.stores` until you find two potential candidates and then choose the one with the better score. This be a power-of-two-choices style selection (https://www.eecs.harvard.edu/~michaelm/postscripts/mythesis.pdf).
</details>
There was an another PR (that might have been reverted too) that exposed this data in the UI. Made it really easy to see exactly how constraints would be applied to the stores.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252)*
<!-- Sent from Reviewable.io -->

BramGruneir committed:
storage: update output for allocator_test's Example_rebalancing

BramGruneir commented:
Fixed a number of the issues.  Haven't added the env variable yet.  That will be next.

---

Review status: 9 of 14 files reviewed at latest revision, 18 unresolved discussions.

---

*[pkg/storage/allocator.go, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5B5X-r5F4buO3kwzP:-KV6E-LAJesfW12HUV9g:b-ldnvj9) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L115)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this can be a value
</details>
Done.  And I also changed the rule solver's functions to have value receivers.

---

*[pkg/storage/allocator.go, line 185 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5BLBfMmoWhW7y4Rpc:-KV6hfvoceV9zUEMavTu:bpy6yh7) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L185)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Why is deterministic order "Very Bad"? Isn't the whole idea of ranking by score such that you can return the best possible candidate? If you're randomizing doesn't that defeat the whole point of positive vs required attributes?
> 
> Isn't it being deterministic a net win since it's then also much easier to understand what's happening?
</details>
I've added this to the follow up issue.

---

*[pkg/storage/allocator.go, line 201 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5BYIA_c5GRHSQhXkl:-KV6GnJyT_PC0ZrAETY4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L201)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> expand this rationale; why is it currently required, why might it not be required? why is it OK to use a bogus range ID?
</details>
Done.

---

*[pkg/storage/allocator.go, line 207 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5ElRI1pw1kbbooWAz:-KV6HbgDjh5IKWhZL2LJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L207)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> seems like you can initialize this to `math.MaxFloat64` and remove `found`.
</details>
Done.

---

*[pkg/storage/allocator.go, line 217 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5EiAqh5ZTwp0j2l7z:-KV6H3l5tjSt2adKHaHd:bckv633) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L217)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why is it OK to skip checking the error?
</details>
not an error being returned, a bool that is not required since we're looking for the worst candidate and that may be `0, false`

---

*[pkg/storage/allocator.go, line 269 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5F276RijvKQ6OxzcR:-KV6I4vEK2cG4wmgn1C1:b-ybqzdl) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L269)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Passing in the storePool was done to support viewing this data in the UI. Since getStoreList wasn't exported, the easiest thing was passing in storePool which was accessible from the admin RPC endpoints.
> 
> https://github.com/cockroachdb/cockroach/pull/8887/commits/4601e629a69483721dea1bbe0d46eb909f9e690f 
</details>
Much cleaner.  And a simple fix, I like it. Done.

I'm going to work on that part in the future.  Might just export GetStoreList if needed.  Added to the follow up issues.

---

*[pkg/storage/allocator.go, line 299 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5GLJXHMbI5kAoE-RY:-KV6dA_cwDfAV-W7ihRg:b-zavg6m) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L299)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I had to squint to figure out what this counter was doing. How about this instead:
> ```
> if len(candidates) > len(existing) {
>   candidates = candidates[:len(existing)]
> }
> ```
</details>
This whole check was very ugly.  I've re-written it to be more explicit.

---

*[pkg/storage/allocator.go, line 310 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5YHbU2nVulMoI0Z-s:-KV6iR1k7_0Gfdbrglal:b-w2l929) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L310)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> We need randomness here too. See my other comment about the candidates being sorted by score at this point making this loop deterministic.
</details>
Yep.  I'll address this in a follow up PR.

---

*[pkg/storage/allocator_test.go, line 674 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5Gp3dZhRSUOEkAgzy:-KV6MFod_aIpZLEFYIsU:beluqw6) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator_test.go#L674)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> doesn't this want to be inside `createTestAllocator`?
</details>
not all tests want a deterministic store pool.

---

*[pkg/storage/allocator_test.go, line 1435 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5H1PiQoAWvQpdl15m:-KV6dPHHxxI2lWkD7VTh:b-195j5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator_test.go#L1435)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> please revert this, or put it in a separate commit. This test is important for validating that the allocator's behaviour didn't inadvertently change, and including this graphical change in the rest of the change makes it difficult to spot interesting changes.
</details>
Sure, I can split it into a second commit.

---

*[pkg/storage/client_raft_test.go, line 1917 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5Dub6Ahg1EL0Mcodz:-KV6fJDdvfXXVF9P2y3W:b-ywobxi) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/client_raft_test.go#L1917)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why'd this change? the diff alone isn't enough?
</details>
Not sure, was part of the change.  I'll just revert it.

---

*[pkg/storage/rule_solver.go, line 81 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV6YeXjELWyKGtR5kbO:-KV6fZvFpvHHRYkEcQcf:b-cf9ljw) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/rule_solver.go#L81)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> The generality of a configurable list of rules feels like overkill to me.
</details>
I agree.  But I didn't want to make too many structural changes to this PR.  I'll put in a todo to cut this down in the future.

---

*[pkg/storage/rule_solver.go, line 122 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV6XcTqVlI3w-r8p9ok:-KV6fvg2UoNpvwwEOX1z:b22k1ch) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/rule_solver.go#L122)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> There was an another PR (that might have been reverted too) that exposed this data in the UI. Made it really easy to see exactly how constraints would be applied to the stores.
</details>
Yeah, power of 2 is bar far the easiest way to go here and I was thinking of something along the same lines.  Added as a todo and to the follow up issue..

---

*[pkg/storage/rule_solver.go, line 127 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV6X4xn8aWF55Hy2iyl:-KV6iUkajvcEr3h8BZT5:b-u8yr5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/rule_solver.go#L127)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> It is super confusing that `candidate` is a type and we see `candidate, bool` here while in `ruleReplicasUniqueNodes` we see `candidate bool`. 
</details>
Indeed.  Fixing.

---

*[pkg/storage/store_pool.go, line 216 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5DUXc1K6Piy_gSsh1:-KV6MY2Ib06gCmBpMU40:b-nubtx8) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/store_pool.go#L216)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this doesn't need to be under lock, does it?
</details>
It's used in getStoreList which needs to be thread safe.

---

*[pkg/storage/store_pool.go, line 518 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5WqWBDmucaEvwgmuK:-KV6npMaUforHXwTEIKi:b-s4sqsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/store_pool.go#L518)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Relying on the map iteration to randomize the `StoreList` is subtle. We should shuffle if `!sp.mu.deterministic`. See `ReplicaSlice.Suffle`.
</details>
Good idea.  I'll address this in a follow up.

---

*[pkg/storage/store_pool_test.go, line 42 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5DeBlR1A5TGBlC3xj:-KV6MolHSuf7KFs_uwTF:bb8dvm9) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/store_pool_test.go#L42)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this has one caller.
</details>
Two now.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-:-KV7-1vC4qrhMZBQleU5:b-emxdsl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 9 of 14 files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[pkg/storage/allocator.go, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5B5X-r5F4buO3kwzP:-KV70qCO51smfAa2Jt87:bxxi00l) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L115)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Done.  And I also changed the rule solver's functions to have value receivers.
</details>
Changing this field to a value is fine, but I object to the knee jerk to use value receivers whenever possible. Doing so costs stack space especially when used pervasively. More stack space means more stack growth which is not free.

---

*[pkg/storage/allocator.go, line 185 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5BLBfMmoWhW7y4Rpc:-KV702ML1f7hBApoQzEy:b30vso4) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L185)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> I've added this to the follow up issue.
</details>
@d4l3k https://brooker.co.za/blog/2012/01/17/two-random.html. The allocator doesn't have perfect knowledge of the current system or have knowledge of what decisions other nodes are making. If multiple nodes are all selecting the same "best" candidate, that candidate will quickly become overloaded. Randomization doesn't mean tossing out constraints, but applying randomness within the constraints. For example, we select a set of candidates that match the required and positive attributes. If that set is empty, we drop the positive attributes. From that list we select 2 candidates and pick the one with the best score.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252)*
<!-- Sent from Reviewable.io -->

tamird commented:
Please move the second commit into a separate PR altogether. I would like to see this PR go in with no change in the output of `Example_rebalancing`.

---

Reviewed 3 of 5 files at r2.
Review status: 11 of 14 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[pkg/storage/allocator.go, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5B5X-r5F4buO3kwzP:-KV70mtqamj4YdAq-Rsi:bsec6e5) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/allocator.go#L115)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Done.  And I also changed the rule solver's functions to have value receivers.
</details>
FYI: there is no need to make the rule solver's method have value receivers - the Allocator is almost certainly on the heap already, and so taking the address of one of its members is likely allocation-free. Not saying you should change anything, just pointing this out.

---

*[pkg/storage/allocator.go, line 213 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV714u0niyTj9GeQnZH:-KV714u0niyTj9GeQnZI:b-clzyoe) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ef40f323b94005dd84378669da8296c9fec8e7c/pkg/storage/allocator.go#L213)):*
> ```Go
> 	// requires a store list, but we should be able to create one using only
> 	// the stores that belong to the range.
> 	// Use an invalid range ID as we don't care about a corrupt replicas since
> ```

"since as"?

---

*[pkg/storage/allocator.go, line 334 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV71nkKmZMeh8jjzKD2:-KV71nkKmZMeh8jjzKD3:btnynrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ef40f323b94005dd84378669da8296c9fec8e7c/pkg/storage/allocator.go#L334)):*
> ```Go
> 
> 	// TODO(bram): #10275 Need some randomness here!
> 	for _, cand := range candidates {
> ```

the candidates are sorted.

---

*[pkg/storage/allocator_test.go, line 1437 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KVAjOQM_79N4KQRpOKD:-KVAjOQM_79N4KQRpOKE:b-y3wh4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ef40f323b94005dd84378669da8296c9fec8e7c/pkg/storage/allocator_test.go#L1437)):*
> ```Go
> 
> 	// Output:
> 	// 999 129 000 000 000 000 000 000 000 000 000 000 000 000 000 000 000 000 000 000
> ```

this is alarming. why did this change?

---

*[pkg/storage/store_pool.go, line 216 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-KV5DUXc1K6Piy_gSsh1:-KVAjWVxaBZW_LDe5qQ4:b-1dzv92) ([raw file](https://github.com/cockroachdb/cockroach/blob/49a994bf245945f73774d9b785fc3507cc6e2a92/pkg/storage/store_pool.go#L216)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> It's used in getStoreList which needs to be thread safe.
</details>
What? I mean that it shouldn't really be modified after creation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10252#-:-KVAjfPerdmPcwL7g0lG:b-ulo8qp)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Closing this PR and I'll be reapplying the rule solver is a collection of smaller PRs.
